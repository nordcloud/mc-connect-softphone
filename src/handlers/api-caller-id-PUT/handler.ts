import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import { StatusCodes } from 'http-status-codes';

import { getUser } from '../../utils/getUser';
import { putCallerId } from './putCallerId';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return { statusCode: StatusCodes.UNAUTHORIZED };
  }

  const callerId = event.body || '';
  const { expiresTimestampSeconds } = await putCallerId({ user, callerId });

  return {
    statusCode: StatusCodes.OK,
    body: JSON.stringify({ expiresTimestampSeconds }),
  };
};
