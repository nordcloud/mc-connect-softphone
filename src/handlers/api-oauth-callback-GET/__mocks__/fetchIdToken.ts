export const fetchIdToken = jest.fn().mockResolvedValue({
  id_token: 'dummyIdToken',
  expires_in: 123456,
});
