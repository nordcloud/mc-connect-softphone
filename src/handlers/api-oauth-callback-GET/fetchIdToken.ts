import { APIGatewayProxyEventV2 } from 'aws-lambda';
import fetch from 'node-fetch';
import { JsonValue } from 'type-fest';

import { LOGIN_FLOW_COOKIE } from '../../consts';
import { OauthAuthorizeResponse } from '../../types';
import { getCookie } from '../../utils/getCookie';
import { getIdTokenPayload } from '../../utils/getIdTokenPayload';
import { getStageConsts } from '../../utils/getStageConsts';
import { assertLoginFlowCookie } from './assertLoginFlowCookie';
import { checkOAuthUrlParams } from './checkOAuthUrlParams';

const { OAUTH_BASE_URL, OAUTH_CLIENT_ID } = getStageConsts();

export async function fetchIdToken(event: APIGatewayProxyEventV2) {
  const { LOGIN_CALLABCK_URL } = process.env;

  if (!LOGIN_CALLABCK_URL) {
    throw new Error('Missing LOGIN_CALLABCK_URL');
  }

  if (!checkOAuthUrlParams(event.queryStringParameters)) {
    console.log('Invalid OAuth callback url params');
    return undefined;
  }

  const { state, code } = event.queryStringParameters;

  const loginFlowCookieJSON = getCookie(event, LOGIN_FLOW_COOKIE);

  if (!loginFlowCookieJSON) {
    console.log('Missing loginFlowCookie');
    return undefined;
  }

  const loginFlowCookie = JSON.parse(loginFlowCookieJSON) as JsonValue;

  assertLoginFlowCookie(loginFlowCookie);

  if (state !== loginFlowCookie.stateNonce) {
    console.log('State nonce mismatch');
    return undefined;
  }

  console.log('Fetching ID token');

  const res = await fetch(`${OAUTH_BASE_URL}/token`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      code,
      code_verifier: loginFlowCookie.codeVerifier,
      grant_type: 'authorization_code',
      redirect_uri: LOGIN_CALLABCK_URL,
      client_id: OAUTH_CLIENT_ID,
    }),
  });

  const data = (await res.json()) as OauthAuthorizeResponse;

  if ('error' in data) {
    throw new Error(`${data.error}: ${data.error_description}`);
  }

  const { nonce } = getIdTokenPayload(data.id_token);

  if (nonce !== loginFlowCookie.idTokenNonce) {
    console.log('ID token nonce mismatch');
    return undefined;
  }

  return data;
}
