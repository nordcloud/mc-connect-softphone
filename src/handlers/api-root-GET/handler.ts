import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { deleteCallerId } from '../../utils/deleteCallerId';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  await deleteCallerId(user.email);

  const customers = await getCustomers();

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers,
  });
};
