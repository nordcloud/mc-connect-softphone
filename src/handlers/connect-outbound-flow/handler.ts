import { ConnectContactFlowHandler } from 'aws-lambda';

import { deleteCallerId } from '../../utils/deleteCallerId';

export const handler: ConnectContactFlowHandler = async (event) => {
  const username = event.Details.Parameters.Username;

  if (!username) {
    throw new Error('Missing Username');
  }

  const callerId = await deleteCallerId(username);

  return {
    caller_id: callerId || null,
  };
};
