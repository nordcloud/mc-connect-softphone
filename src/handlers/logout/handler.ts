import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getCookie } from '../../utils/getCookie';
import { getStageConsts } from '../../utils/getStageConsts';

const { OAUTH_BASE_URL } = getStageConsts();

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const { API_URL } = process.env;

  if (!API_URL) {
    throw new Error('Missing API_URL');
  }

  const idToken = getCookie(event, ID_TOKEN_COOKIE);
  let location = `${OAUTH_BASE_URL}/logout?id_token_hint=${idToken}&post_logout_redirect_uri=${API_URL}`;

  if (!idToken) {
    console.log('Missing idToken cookie');
    location = '/';
  }

  console.log(`Redirecting to ${location}`);

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location },
    cookies: [cookie.serialize(ID_TOKEN_COOKIE, '')],
  };
};
