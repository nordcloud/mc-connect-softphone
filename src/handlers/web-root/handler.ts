import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { AWS_CONNECT_URL } from '../../consts';
import { getCustomers } from '../../utils/getCustomers';
import { getUser } from '../../utils/getUser';
import { loginRedirect } from '../../utils/loginRedirect';
import { renderHTML } from '../../utils/renderHTML';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const user = await getUser(event);

  if (!user) {
    return loginRedirect();
  }

  return renderHTML('homepage', {
    AWS_CONNECT_URL,
    user,
    customers: await getCustomers(event),
  });
};
