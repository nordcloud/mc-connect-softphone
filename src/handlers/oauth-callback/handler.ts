import { APIGatewayProxyHandlerV2 } from 'aws-lambda';
import cookie from 'cookie';
import { StatusCodes } from 'http-status-codes';

import { ID_TOKEN_COOKIE } from '../../consts';
import { getStage } from '../../utils/getStage';
import { fetchIdToken } from './fetchIdToken';

export const handler: APIGatewayProxyHandlerV2 = async (event) => {
  const responseCookies = [];
  const idTokenData = await fetchIdToken(event);

  if (idTokenData) {
    responseCookies.push(
      cookie.serialize(ID_TOKEN_COOKIE, idTokenData.id_token, {
        secure: getStage() !== 'local',
        httpOnly: true,
        expires: new Date(Date.now() + idTokenData.expires_in * 1000),
      })
    );
  }

  console.log('Redirecting to "/"');

  return {
    statusCode: StatusCodes.MOVED_TEMPORARILY,
    headers: { location: '/' },
    cookies: responseCookies,
  };
};
