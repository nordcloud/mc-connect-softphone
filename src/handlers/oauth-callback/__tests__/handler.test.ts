import cookie from 'cookie';

import { getStage } from '../../../utils/getStage';
import { fetchIdToken } from '../fetchIdToken';
import { handler } from '../handler';

jest.mock('../fetchIdToken');
jest.mock('../../../utils/getStage');

const args = ['dummyEvent'] as unknown as Parameters<typeof handler>;

describe('oauth-callback handler', () => {
  it('calls fetchIdToken with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(fetchIdToken).mock.calls).toMatchSnapshot();
  });

  it('calls cookie.serialize with correct args', async () => {
    await handler(...args);
    expect(jest.mocked(cookie.serialize).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });

  describe('when stage is "local"', () => {
    beforeEach(() => {
      jest.mocked(getStage).mockReturnValue('local');
    });

    it('calls cookie.serialize with correct args', async () => {
      await handler(...args);
      expect(jest.mocked(cookie.serialize).mock.calls).toMatchSnapshot();
    });
  });

  describe('when idTokenData is missing', () => {
    beforeEach(() => {
      jest.mocked(fetchIdToken).mockResolvedValueOnce(undefined);
    });

    it('resolves with a correct value', () => {
      return expect(handler(...args)).resolves.toMatchSnapshot();
    });
  });
});
