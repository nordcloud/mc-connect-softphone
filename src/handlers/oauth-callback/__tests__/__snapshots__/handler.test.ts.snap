// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`oauth-callback handler calls cookie.serialize with correct args 1`] = `
[
  [
    "idToken",
    "dummyIdToken",
    {
      "expires": 2023-03-24T00:40:16.529Z,
      "httpOnly": true,
      "secure": true,
    },
  ],
]
`;

exports[`oauth-callback handler calls fetchIdToken with correct args 1`] = `
[
  [
    "dummyEvent",
  ],
]
`;

exports[`oauth-callback handler resolves with a correct value 1`] = `
{
  "cookies": [
    "dummyCookieSerializeValue",
  ],
  "headers": {
    "location": "/",
  },
  "statusCode": 302,
}
`;

exports[`oauth-callback handler when idTokenData is missing resolves with a correct value 1`] = `
{
  "cookies": [],
  "headers": {
    "location": "/",
  },
  "statusCode": 302,
}
`;

exports[`oauth-callback handler when stage is "local" calls cookie.serialize with correct args 1`] = `
[
  [
    "idToken",
    "dummyIdToken",
    {
      "expires": 2023-03-24T00:40:16.529Z,
      "httpOnly": true,
      "secure": false,
    },
  ],
]
`;
